File: 15. 3Sum/15. 3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b);
    var res = [];
    
    for (let i = 0; i < nums.length - 2; i++) {
        if (nums[i] > 0) {
            break;
        };
        
        if (i > 0 && nums[i] === nums[i-1]) {
            continue;
        };
        
        let j = i+1, k = nums.length - 1;
        while (j < k) {
            let total = nums[i] + nums[j] + nums[k];
            if (total < 0) {
                j++;
            }
            else if (total > 0) {
                k--;
            }
            else {
                res.push([nums[i], nums[j], nums[k]]);
                while (j < k && nums[j] == nums[j+1]) {
                    j++;
                }
                j++;
                
                while (j < k && nums[k] == nums[k-1]) {
                    k--;
                }
                k--;
            }
        }
    }
    return res;
};

File: 15. 3Sum/15. 3Sum.py
class Solution:
    def threeSum(self, nums: 'List[int]') -> 'List[List[int]]':
        total_nums = len(nums)
        if total_nums == 0:
            return []
        else:
            solutions = []
            nums.sort()
            
            for i in range(total_nums - 2):
                if nums[i] > 0:
                    break
                if i > 0 and nums[i] == nums[i-1]:
                    continue
                j = i + 1
                k = total_nums - 1
                
                while j < k:
                    sums = nums[i] + nums[j] + nums[k]
                    if sums < 0:
                        j += 1
                    elif sums > 0:
                        k -= 1
                    else:
                        solutions.append([nums[i], nums[j], nums[k]])
                        while j < k and nums[j] == nums[j+1]:
                            j += 1
                        j += 1
                        
                        while j < k and nums[k] == nums[k-1]:
                            k -= 1
                        k -= 1
                        
            return solutions
